### PR TITLE
flamegraph: Binary based color stack

### LINKED ISSUE #2145: Flamegraph - coloured stacks based on the binary name
Request from Discord: https://discord.com/channels/877547706334199818/877547706334199821/1042763253849272320

### PR DESCRIPTION
**Pull request by manojVivek**
Resolves #2145.

### CODE REVIEW

**Comment by yomete** [comment]
not sure if it's intentional but the root node appears to have a dark background
image

File: ui/packages/shared/components/src/KeyDownContext/index.tsx [hunk 1]
```diff
@@ -39,7 +54,17 @@ const useIsShiftDown = (): boolean => {
...
+export const useKeyDown = (): KeyDownState => {
+  const context = useContext(KeyDownContext);
+  if (context == null) {
+    return DEFAULT_VALUE;
+  }
+  return context;
 };
 
-export default useIsShiftDown;
+export default KeyDownContext;
```
Comment (yomete) [inline ui/packages/shared/components/src/KeyDownContext/index.tsx:70]: Looks like you forgot to use this context provider. Maybe you can add it to the App.tsx file in the `packages/app/web/src` folder

File: ui/packages/shared/profile/src/ProfileIcicleGraph/IcicleGraph/ColorStackLegend.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import {useMemo} from 'react';
+import {useURLState} from '@parca/components';
+import type {NavigateFunction} from '@parca/functions';
+
+interface Props {
+  navigateTo?: NavigateFunction;
+  compareMode?: boolean;
+}
+
+const ColorStackLegend = ({navigateTo, compareMode = false}: Props): JSX.Element => {
```
Comment (yomete) [inline ui/packages/shared/profile/src/ProfileIcicleGraph/IcicleGraph/ColorStackLegend.tsx:27]: There seems to be some inconsistencies between the spacing of the color legends, and also the `x` icon gets shrunk in some cases.

image

image

We could also try out making the legends start from the beginning of the div.
image

**Review by yomete — APPROVED** [review]
nice work! just a few observations about some design changes and the 'hold shift when hovering' tooltip interaction not working.

File: ui/packages/shared/components/src/KeyDownContext/index.tsx (hunk 1 above)
Comment (manojVivek) [reply ui/packages/shared/components/src/KeyDownContext/index.tsx:70]: Ah, sorry missed the `MetricsGraph`, now wrapped the whole `ProfileExplorer`.  

Trying to minimize the side effect changes on the cloud.

File: ui/packages/shared/profile/src/ProfileIcicleGraph/IcicleGraph/ColorStackLegend.tsx (hunk 2 above)
Comment (manojVivek) [reply ui/packages/shared/profile/src/ProfileIcicleGraph/IcicleGraph/ColorStackLegend.tsx:27]: > There seems to be some inconsistencies between the spacing of the color legends, and also the `x` icon gets shrunk in some cases.

Was using `grid` that made the spacing look uneven, now changed it to `flex` and it looks better.

> We could also try out making the legends start from the beginning of the div.

Fixed this too.

**Comment by manojVivek** [comment]
> not sure if it's intentional but the root node appears to have a dark background

I'm unable to see this behaviour, any idea when're you seeing this?

**Comment by manojVivek** [comment]
Merging this in to get started on #2472. Please feel free to add your comments if any, and I'll take care of them.

**Comment by manojVivek** [comment]
> not sure if it's intentional but the root node appears to have a dark background 

Fixed in https://github.com/parca-dev/parca/pull/2559

**Review by yomete — COMMENTED** [review]
Thanks for addressing the suggestions :)

**Pull request merged**